server/src/main/java/io/crate/planner/optimizer/costs/PlanStats.java

```diff
@@ -106,62 +108,58 @@ public Stats visitGroupReference(GroupReference group, Void context) {
-                var numberOfRows = DataTypes.LONG.sanitizeValue(((Literal<?>) limit.limit()).value());
-                return new Stats(numberOfRows, stats.sizeInBytes(), stats.statsByColumn());
+            if (limit.limit() instanceof Literal<?> literal) {
+                var numberOfRows = DataTypes.LONG.sanitizeValue(literal.value());
+                return stats.withNumDocs(numberOfRows);
             }
             return stats;
         }
 
         @Override
         public Stats visitUnion(Union union, Void context) {
-            var numberOfRows = -1L;
-            var sizeInBytes = -1L;
             var lhsStats = union.lhs().accept(this, context);
             var rhsStats = union.rhs().accept(this, context);
-            if (lhsStats.numDocs() != -1 && rhsStats.numDocs() != -1) {
-                numberOfRows = lhsStats.numDocs() + rhsStats.numDocs();
-            }
-            if (lhsStats.sizeInBytes() != -1 && rhsStats.sizeInBytes() != -1) {
-                sizeInBytes = Math.max(lhsStats.sizeInBytes(), rhsStats.sizeInBytes());
-            }
-            return new Stats(numberOfRows, sizeInBytes, Maps.concat(lhsStats.statsByColumn(), rhsStats.statsByColumn()));
+            return lhsStats.add(rhsStats);
         }
 
         @Override
         public Stats visitNestedLoopJoin(NestedLoopJoin join, Void context) {
-            var numberOfRows = -1L;
-            var sizeInBytes = -1L;
             var lhsStats = join.lhs().accept(this, context);
             var rhsStats = join.rhs().accept(this, context);
-            if (lhsStats.numDocs() != -1 && rhsStats.numDocs() != -1) {
-                if (join.joinType() == JoinType.CROSS) {
-                    numberOfRows = lhsStats.numDocs() * rhsStats.numDocs();
-                } else {
-                    // We don't have any cardinality estimates, so just take the bigger table
-                    numberOfRows = Math.max(lhsStats.numDocs(), rhsStats.numDocs());
-                }
-            }
-            if (lhsStats.sizeInBytes() != -1 && rhsStats.sizeInBytes() != -1) {
-                sizeInBytes = lhsStats.sizeInBytes() + rhsStats.sizeInBytes();
+            Map<ColumnIdent, ColumnStats<?>> statsByColumn = Maps.concat(lhsStats.statsByColumn(), rhsStats.statsByColumn());
+            if (lhsStats.numDocs() == -1
+                || lhsStats.sizeInBytes() == -1
+                || rhsStats.numDocs() == -1
+                || rhsStats.sizeInBytes() == -1) {
+                return new Stats(-1, -1, statsByColumn);
             }
-            return new Stats(numberOfRows, sizeInBytes, Maps.concat(lhsStats.statsByColumn(), rhsStats.statsByColumn()));
+            long numRows = join.joinType() == JoinType.CROSS
+                ? lhsStats.numDocs() * rhsStats.numDocs()
+                // We don't have any cardinality estimates, so just take the bigger table
+                : Math.max(lhsStats.numDocs(), rhsStats.numDocs());
+            return new Stats(
```

Comment:
not part of this PR, just a question: If we have LEFT/SEMI/ANTI shouldn't be `lhsStats.numDocs()` and if RIGHT: `rhsStats.numbDocs()`?

Reply:
Not sure, I think we'll overhaul this logic anyway.